src/tigerbeetle/client.zig

```diff
@@ -0,0 +1,839 @@
+        const Self = @This();
+        const Client = vsr.Client(StateMachine, MessageBus);
+
+        fn fail(comptime fmt: []const u8, args: anytype) noreturn {
+            const stderr = std.io.getStdErr().writer();
+            stderr.print(fmt, args) catch unreachable;
+            std.os.exit(1);
+        }
+
+        fn print(comptime fmt: []const u8, args: anytype) void {
+            const stdout = std.io.getStdOut().writer();
+            stdout.print(fmt, args) catch return;
+        }
+
+        pub const Context = struct {
+            event_loop_done: bool,
+            request_done: bool,
+
+            repl: bool,
+            debug_logs: bool,
+
+            client: *Client,
+            message: ?*MessagePool.Message,
+
+            fn fail(context: *Context, comptime fmt: []const u8, args: anytype) void {
+                if (!context.repl) {
+                    Self.fail(fmt, args);
+                }
+
+                print(fmt, args);
+            }
+
+            fn debug(context: *Context, comptime fmt: []const u8, args: anytype) void {
+                if (context.debug_logs) {
+                    print("[Debug] " ++ fmt, args);
+                }
+            }
+
+            fn fail_at(
+                context: *Context,
+                input: []const u8,
+                index: usize,
+                comptime fmt: []const u8,
+                args: anytype,
+            ) void {
+                var line_number: usize = 1;
+                var column_number: usize = 0;
+
+                var line_beginning: usize = 0;
+                var found_current_line = false;
+
+                var i: usize = 0;
+                while (i < input.len) : (i += 1) {
+                    var c = input[i];
+
+                    if (c == '\n') {
+                        if (!found_current_line) {
+                            line_beginning = i + 1;
+                            line_number += 1;
+                            column_number = 0;
+                        } else {
+                            break;
+                        }
+                    } else {
+                        if (!found_current_line) {
+                            column_number += 1;
+                        }
+                    }
+
+                    if (i == index) {
+                        found_current_line = true;
+                    }
+                }
+
+                var line = input[line_beginning..i];
+
+                const stderr = std.io.getStdErr().writer();
+                stderr.print("Fail near line {}, column {}:\n\n{s}\n", .{
+                    line_number,
+                    column_number,
+                    line,
+                }) catch unreachable;
+                while (column_number > 0) {
+                    stderr.print(" ", .{}) catch unreachable;
+                    column_number -= 1;
+                }
+                stderr.print("^ Near here.\n\n", .{}) catch unreachable;
```

Comment:
I know it would be unlike what we do elsewhere, but the panic is also kind of user unfriendly. If I make a syntax error in other database CLIs, it just tells me the error like this rather than giving me a panic that includes the source code of the database and other info that is irrelevant to my typo?

Reply:
Oh sorry, the printing to stderr is fine, I just mean bubble up instead of using `catch unreachable`.
i.e.:

```
try stderr.print("..");
```

Reply:
Oh gotcha, yep makes sense.